### PR TITLE
token jobject static file route map and Token JObject

### PR DESCRIPTION
**Pull request by qinxiaowen**

### CODE REVIEW

File: src/Andoromeda.Kyubey.Models/KyubeyContext.cs [hunk 1]
```diff
@@ -178,7 +178,7 @@ protected override void OnModelCreating(ModelBuilder builder)
 
             builder.Entity<Token>(e =>
             {
-                e.HasIndex(x => x.Priority);
+                //e.HasIndex(x => x.Priority);
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Models/KyubeyContext.cs]: Don't remove pri in DB.

File: src/Andoromeda.Kyubey.Models/Token.cs [hunk 2]
```diff
@@ -33,12 +33,12 @@ public class Token
...
-        public byte[] Icon { get; set; }
+        //public byte[] Icon { get; set; }
 
         [MaxLength(32)]
         public string Name { get; set; }
 
-        public string Description { get; set; }
+        //public string Description { get; set; }
 
         public long? DeliveryAmount { get; set; }
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Models/Token.cs]: Which place the DeliveryAmount has been used?

File: src/Andoromeda.Kyubey.Portal/Controllers/AnalysisController.cs [hunk 3]
```diff
@@ -5,11 +5,16 @@
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.EntityFrameworkCore;
+using Pomelo.AspNetCore.Localization;
 
 namespace Andoromeda.Kyubey.Portal.Controllers
 {
     public class AnalysisController : BaseController
     {
+        public AnalysisController(ICultureProvider cultureProvider) : base(cultureProvider)
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Controllers/AnalysisController.cs]: Please inject ICultureProvider in BaseController

Comment (yukozh) [reply src/Andoromeda.Kyubey.Portal/Controllers/AnalysisController.cs]: Don't use constructor to inject any services

File: src/Andoromeda.Kyubey.Portal/Controllers/HomeController.cs [hunk 4]
```diff
@@ -9,34 +9,44 @@
...
+        }
         [Route("/")]
         public async Task<IActionResult> Index([FromServices] KyubeyContext db)
         {
-            //linq 待优化
-            var tokens = await db.TokenHatchers
-                .Include(x => x.Token)
-                .Select(x => new TokenHandlerListVM()
+            var tokenStaticInfoList = _tokenRepository.GetAll();
+            var tokens = (await db.TokenHatchers.Where(x => x.Status == Kyubey.Models.Hatcher.TokenHatcherStatus.Active)
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Controllers/HomeController.cs]: Why we still need this table?

File: src/Andoromeda.Kyubey.Portal/Interface/IBaseRepository.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+
+namespace Andoromeda.Kyubey.Portal.Interface
+{
+    public interface IBaseRepository<T>
+    {
+
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Interface/IBaseRepository.cs]: Empty interface?

File: src/Andoromeda.Kyubey.Portal/Services/FileHelper.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Andoromeda.Kyubey.Portal.Services
+{
+    public static class FileHelper
+    {
+        public static void RemoveDangerousChar(this string fileName)
+        {
+            fileName = fileName.Replace(".", "").Replace("\\", "").Replace("/", "");
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Services/FileHelper.cs]: What's this?

File: src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    public class TokenFileInfoRepository : ITokenRepository
+    {
+        private const string tokenFolderRelativePath = @"Tokens";
+        private string tokenFolderAbsolutePath = Path.Combine(System.AppDomain.CurrentDomain.BaseDirectory, tokenFolderRelativePath);
+        private const string manifestFileName = "manifest.json";
+        private const string iconFileName = "icon.png";
+
+        private string GetFileNameSuffixByCulture(string cultureStr)
+        {
+            if (new string[] { "en", "en-US", "en-GB" }.Contains(cultureStr))
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs]: en-US, en-GB are illegal. Only the first string is legal.

File: src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            if (availablePath != null)
+                return FileHelper.ReadAllText(availablePath);
+            return null;
+        }
+
+        public string GetTokenIncubationDetail(string tokenId, string cultureStr)
+        {
+            tokenId.RemoveDangerousChar();
+            var folderPath = Path.Combine(tokenFolderAbsolutePath, tokenId, "incubator");
+            var files = FileHelper.GetAllFileNameFromFolder(folderPath, "detail*.md");
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs]: `detail.*.md`?

File: src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            var availableFiles = GetAvailableFileNames(files, cultureStr);
+            var availablePath = availableFiles.Select(x => Path.Combine(folderPath, x)).FirstOrDefault();
+            if (availablePath != null)
+                return FileHelper.ReadAllText(availablePath);
+            return null;
+        }
+
+        //public string GetTokenDescription(string tokenId, string cultureStr)
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs]: Please don't check in any commented codes.

File: src/Andoromeda.Kyubey.Portal/Services/TokenTool.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    public class TokenTool
+    {
+        public static string GetTokenIncubatorBannerUri(string tokenId, string filePath)
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                return Path.Combine("img", "null.png");
+            }
+            string fileName = Path.GetFileName(filePath);
+            return $"/Token/tokenbanner/{tokenId}/{fileName}";
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Services/TokenTool.cs]: Why return this? Did you mapped the token_assets?

File: src/Andoromeda.Kyubey.Portal/Views/Home/Index.cshtml [hunk 11]
```diff
@@ -1,4 +1,4 @@
-﻿@model IEnumerable<TokenHandlerListVM>
+﻿@model IEnumerable<TokenIncubatorListVM>
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Views/Home/Index.cshtml]: Don't use abbreviate

File: src/Andoromeda.Kyubey.Portal/Views/Token/_TokenInfoBox.cshtml [hunk 12]
```diff
@@ -16,7 +17,7 @@
                     </div>
                     <div class="token-incubation-description-outer">
                         <h3>@Model.Id</h3>
-                        <div style="">@Model.Description</div>
+                        <div>@ViewBag.TokenDescription</div>
```
Comment (yukozh) [inline src/Andoromeda.Kyubey.Portal/Views/Token/_TokenInfoBox.cshtml]: Did you rendered markdown? Please use @Html.Marked(ViewBag.TokenDescription)

File: src/Andoromeda.Kyubey.Models/KyubeyContext.cs (hunk 1 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Models/KyubeyContext.cs]: done

File: src/Andoromeda.Kyubey.Models/Token.cs (hunk 2 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Models/Token.cs]: This field is still in use

File: src/Andoromeda.Kyubey.Portal/Controllers/AnalysisController.cs (hunk 3 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Controllers/AnalysisController.cs]: done

File: src/Andoromeda.Kyubey.Portal/Controllers/HomeController.cs (hunk 4 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Controllers/HomeController.cs]: removed

File: src/Andoromeda.Kyubey.Portal/Interface/IBaseRepository.cs (hunk 5 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Interface/IBaseRepository.cs]: removed

File: src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs (hunk 8 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs]: done

File: src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs (hunk 9 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Services/TokenFileInfoRepository.cs]: done

File: src/Andoromeda.Kyubey.Portal/Views/Home/Index.cshtml (hunk 11 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Views/Home/Index.cshtml]: done

File: src/Andoromeda.Kyubey.Portal/Views/Token/_TokenInfoBox.cshtml (hunk 12 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Views/Token/_TokenInfoBox.cshtml]: it's not markdwon text

File: src/Andoromeda.Kyubey.Portal/Services/TokenTool.cs (hunk 10 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Services/TokenTool.cs]: improve it later

File: src/Andoromeda.Kyubey.Portal/Services/FileHelper.cs (hunk 6 above)
Comment (qinxiaowen) [reply src/Andoromeda.Kyubey.Portal/Services/FileHelper.cs]: this is filter illegal string like this "../../../windows/"

**Pull request merged**